src/network/operator.rs

```diff
@@ -0,0 +1,402 @@
+                            operator.known_nonces.write().await.clear();
+                        }
+
+                        // Sleep for `5` seconds.
+                        tokio::time::sleep(UPDATE_DELAY).await;
+                    }
+                }));
+                // Wait until the operator handler is ready.
+                let _ = handler.await;
+            } else {
+                error!("Missing operator address. Please specify an Aleo address in order to operate a pool");
+            }
+        }
+
+        Ok(operator)
+    }
+
+    /// Returns an instance of the operator router.
+    pub fn router(&self) -> OperatorRouter<N> {
+        self.operator_router.clone()
+    }
+
+    /// Returns all the shares in storage.
+    pub fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.state.to_shares()
+    }
+
+    ///
+    /// Performs the given `request` to the operator.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: OperatorRequest<N>) {
+        match request {
+            OperatorRequest::PoolRegister(peer_ip, address) => {
+                if let Some(block_template) = self.block_template.read().await.clone() {
+                    // Ensure this prover exists in the list first, and retrieve their share difficulty.
+                    let share_difficulty = self
+                        .provers
+                        .write()
+                        .await
+                        .entry(address)
+                        .or_insert((Instant::now(), BASE_SHARE_DIFFICULTY, 0))
+                        .1;
+
+                    // Route a `PoolRequest` to the peer.
+                    let message = Message::PoolRequest(share_difficulty, Data::Object(block_template));
+                    if let Err(error) = self.peers_router.send(PeersRequest::MessageSend(peer_ip, message)).await {
+                        warn!("[PoolRequest] {}", error);
+                    }
+                } else {
+                    warn!("[PoolRegister] No current block template exists");
+                }
+            }
+            OperatorRequest::PoolResponse(peer_ip, block_header, prover_address) => {
+                if let Some(block_template) = self.block_template.read().await.clone() {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block_header.height() {
+                        warn!("[PoolResponse] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Ensure the nonce hasn't been seen before.
+                    if self.known_nonces.read().await.contains(&block_header.nonce()) {
+                        warn!("[PoolResponse] Peer {} sent a duplicate share", peer_ip);
+                        // TODO (julesdesmit): punish?
```

Comment:
Just making sure this doesn't need to be addressed yet

Reply:
I thought about it, but it makes more sense to tackle this from the P2P part of the system by implementing blacklisting logic or something like that. I don't think that logic really belongs here.